### ISSUE TITLE
Improvements for y4

### ISSUE THREAD

**Issue by marc-hanheide** [strands-project/aaf_deployment#344]
Please put all the minor improvements here:
- [x] bigger clock
- [ ] show an alternative screen between tasks
- [x] make sure tasks are not dropped
- [x] make it really available to end-users on request
- [ ] thinks about ways to give more control to end-users.
- [x] make bellbot available on demand at various places

**Comment by cdondrup** [comment]
https://github.com/strands-project/aaf_deployment/blob/indigo-devel/voice_output/scripts/task_monitor.py could be adjusted to not just say stuff when a task at a different waypoint is started but also show a screen once the robot drives off.

Edit: that could also show the arrows indicating directions.

**Comment by TobKoer** [comment]
@marc-hanheide 
- [ ] fonts should be bigger for older adults
- [ ] make clock bigger or better make it also an icon that can be touched and then opens a bigger clock screen
- [ ] show alternative screen between tasks (to make it clear that robot cannot be stopped when it's moving for info-terminal or bellbot!)
- [ ] enable users to choose between info-terminal function and bellbot function on screen --> make bellbot more interactive; enable users (esp. visitors) to trigger bellbot themselves.
- [ ] make info-terminal interface more usable for elderly residents. (e.g. the resident-menu icon was frequently overlooked by users, back buttons also sometimes, we suggest making the menu for all info-terminal options available on each level; e.g. menu bar that's always available)
- [ ] add cognitive training or cognitively stimulating game (visible and usable for dementia patients; German language; easy levels or adaptive difficulty!)
- [ ] improve screen sensitivity for touching
- [ ] automatise food-menu update

**Comment by denisehe** [comment]
Yes everything I can think of streigth away, is already listed - maybe more details are coming after analysis.
And I think how to procede in terms of walking group should be discussed anyways during the meeting in the UK.

**Comment by TobKoer** [comment]
@marc-hanheide 
Here's more detailed feedback from our info-terminal assessments with older residents (by kind permission of M. Weninger):
-   Missing feedback: long loading times require feedback so that users know that the touchscreen has registered their action
  o   There should be appropriate form of feedback that the tap of the user was registered and the new page is already loading
-   Most of our participants never used a computer before and therefore don’t have a concept of a typical menu structure.
  o   Due to the low complexity of the robots GUI, its interface should be redesigned in a way, which makes it possible to provide support for navigation. Environmental support in form of an additional menu bar reduces demands on working memory and facilitates recognition (instead of recall).
-   Some interface elements have not been clear for the residents of the care home. Due to the photos-icon, some thought that they could take pictures with the robot. Furthermore the info-icon and its label wasn’t 100% clear to them.  the presentation of information, content and available user action should be understandable and effective throughout the whole GUI,  regardless of user’s experience, knowledge and level of concentration: each interactive element should be perceived and be understandable, without the need of documentation. The text for important interface elements should be self-explanatory
  o   Replace the current photos-icon by another without a camera on it and lable it with “photo album”. An icon would be fine, which signals the users that they can view pictures (of a gallery).
  o   Lable the information-icon with “infos about Henry”
- Opportunity: if there’s a menu bar on all screens, the clock-menu should be the default start screen.

**Comment by cdondrup** [comment]
Even though I won't be involved much any more but still just a few thoughts:
- Feedback when a button is pressed could be similar to the walking group, i.e. playing a quick jingle. The loading times are absolutely random and I still don't know why it sometimes takes a minute to load a page that normally loads in a second. Another alternative or in addition to the sound would be to use the new Message box but I don't remember if this would vanish when a new page loads.
- Regarding adding a game, the usual answer is that we have nobody working on topics like that in the project. So if there is nothing out of the box that could be used or easily adapted then it's hard to find the time for that. Do you have a suggestion for a simple game?
- Isn't adding a menu to the info terminal also adding more complexity that we didn't want? But if that is the feedback, so be it.

I am happy to help with some of the issues in my spare time but cannot promise anything obviously. I think we should have chat at the review what the exact plans for Y4 are and who will be doing what.

**Comment by marc-hanheide** [comment]
Most sensible contents I've read after browsing the news. @cdondrup for PM. Seriously though, let's have discussion in MK.

**Comment by denisehe** [comment]
@capamat-aaf  see here also open issues for the upcoming deployment!

**Comment by capamat-aaf** [comment]
@marc-hanheide 
Quick, big question, we were at the TU yesterday and I finally got to meet Henry, which, of course, was a pleasure.
I wanted to inquire about the Walking Group task; even though I'm not sure if I want to evaluate them again separately, I hope that the task will not be dropped and will still be available for the people here to use.
Are there any developments or news in this regard?

**Comment by cdondrup** [comment]
I am happy to support the walking group but cannot implement any new features. So it would like it was last deployment.

**Comment by capamat-aaf** [comment]
@cdondrup 
that is great! Thank you for your availability. The procedure will be like in the last deployment; hopefully one minor change can be implemented, but Denise started a new thread on that.

**Comment by Jailander** [comment]
Hi this is the topological map as it was last deployment the area within the yellow square is where the robot was allowed to record images ![aaf_deployment16](https://cloud.githubusercontent.com/assets/5035967/18516584/7c181c56-7a91-11e6-8593-a73ba12614f8.png)

**Comment by Jailander** [comment]
Hi I created some screenshots from the info terminal, this is the current layout
![infot1](https://cloud.githubusercontent.com/assets/5035967/18586395/4da2e962-7c14-11e6-96ac-8a4a490f1789.png)
![infot2](https://cloud.githubusercontent.com/assets/5035967/18586396/4da392cc-7c14-11e6-851c-cdf561744348.png)
![infot3](https://cloud.githubusercontent.com/assets/5035967/18586394/4da03bd6-7c14-11e6-9ce8-be84cdc85df4.png)
![infot4](https://cloud.githubusercontent.com/assets/5035967/18586393/4d9ea1e0-7c14-11e6-80f6-9adac8b136a6.png)
![infot5](https://cloud.githubusercontent.com/assets/5035967/18586392/4d9e9e84-7c14-11e6-94ca-b147d945e250.png)

**Comment by cdondrup** [comment]
AFAIK, the start screen had the video playing in the bottom centre under clock in the latest version. But I have no idea where those changes are.

**Comment by Jailander** [comment]
I was hoping no one would notice :p but yes @cdondrup is right I just didn't run the latest version as it was not in linda

**Comment by capamat-aaf** [comment]
@Jailander thank you for the map and the screenshots!
I will work on mockups for the infoterminal; I googled the screen of the robot and found out that the size is 15 inch but I couldn't find the resolution listed. I assume it is 1024x768, or is it listed in some documentation?

**Comment by marc-hanheide** [comment]
@capamat-aaf yes the resolution is only 1024x768, just like the screen shots @Jailander sent.

**Comment by capamat-aaf** [comment]
@marc-hanheide @Jailander 
Hi,
https://github.com/strands-project/aaf_deployment/issues/348

I uploaded drafts for the infoscreen in this issue, and here's a screenshot:
![info-screen-start](https://cloud.githubusercontent.com/assets/21201058/18834755/ffe8cc28-83f7-11e6-912d-5c545cc97f07.png)

**Comment by capamat-aaf** [comment]
@Jailander could you send me the list for the bellbot or tell me how I can help you update?

**Comment by capamat-aaf** [comment]
So I have had some reports about Henry stepping over people's toes (n=2).
I have been told that they aren't "mapped" so basically there is not much we can do. 
I'll see if this is a reoccurring issue.
(I stood in it's path at one point (corridor in front of AAF) to try it out and Henry would not make it past me, as he stopped at some point next to me and would not  continue)

**Comment by marc-hanheide** [comment]
Is that n=2 toes or people ;-)

Seriously, we are having some problems with the human-aware navigation that have not had ever before, known as the dynamic_reconfigure bug by @jailander and @cdondrup. Not sure if this is in any way related, though. It might be worthwhile asking @kiruwang to spend a day monitoring the robot to watch out for things like misbehaving navigation...

**Comment by capamat-aaf** [comment]
People, unknown toes :)
Thank you! hopefully it will be fixed with those issues and I'll ask @kirumang next time he is here

**Comment by kirumang** [comment]
In my opinion, it is pretty okay when Henry directly moves toward people in front of him. He can safely stop in front of people. And I think that navigation functions don't have serious problem in the normal situations.
But, when he turned himself to avoid people or rotated to check his surroundings,  his tail (where it has a hole for the power connection) sometimes reached to people's toes or wheelchairs closer. 
I think this was caused by its longest distance from the robot's moving center. 
@capamat-aaf  let me know when stepping over toes happen again. I'll go to AAF to check it.

**Comment by capamat-aaf** [comment]
General question: assuming someone is clicking through when Henry is in infoterminal mode and he has a different task coming up or has to move: does he wait until that person is gone?
Someone, who this happened to, inquired on this.

**Comment by marc-hanheide** [comment]
This relates to the scheduler @hawesie. The info terminal sets the is_interuptable to false for 30 second after every click. So it should no happen IF  the scheduler respects that flag.

**Comment by hawesie** [comment]
Ah. Needs a fix. Will do it now.

**Comment by hawesie** [comment]
Fix added to https://github.com/strands-project/strands_executive/pull/282

**Comment by Jailander** [comment]
should I add this into werner?

**Comment by hawesie** [comment]
Yes please.

**Comment by capamat-aaf** [comment]
Thank you!

**Comment by capamat-aaf** [comment]
For protocol purpose: 
We received a letter in the complaint box regarding Henry: on Saturday, someone was rushed in with an emergency and the robot was in the way. 
I assume he got stuck, because there was a lot of commotion around him.
Anyway, people tried to move him, which of course only worked partially, as he wasn't in Free Move Mode.
I guess it was bad luck that it happened on the weekend, otherwise I could've intervened. 
To make sure: how can I set it in free-move-mode? The button on his side, iirc. 
Or should I use the Emergency Stop on the control UI in such cases, fire alarms etc?

**Comment by marc-hanheide** [comment]
:shit: ok I suggest we implement a little button in the GUI that engages free run for 30 second. That carries the risk that people click it by accident but it is still better than this!! We need to make sure the robot is not a risk to anyone or well get into trouble.

**Comment by capamat-aaf** [comment]
That's neat. There is even a free spot in the menu bar.

**Comment by marc-hanheide** [comment]
Well... we can't have it only in the info-terminal... It needs to be present in all screens, as the robot might be doing walking group or something else at the moment. So I was planning on a small red cross or something in the top right corner hovering over any interface.

**Comment by Jailander** [comment]
I think @gestom 's virtual bumper works, it would just need testing (and finding it) I guess. The small red cross makes sense but maybe we would need some training for people to know what it means or a sign on top of the screen?

**Comment by capamat-aaf** [comment]
True, but if bottom right is also an option, then it could go in the menu bar and thus hopefully also look somewhat better.

**Comment by cdondrup** [comment]
In the bottom right it would overlay the "back" button in the walking group interface.

Also, what would that button do? Would it trigger the emergency stop or just turn on freerun? If the latter, keep in mind that as soon as the robot receives a new nav goal, it would disengage freerun again.

**Comment by capamat-aaf** [comment]
Can one pause the new nav goal list for the 30s?
But maybe that is not a problem, if it's pushed out of the way, it can restart routine somewhere else then. (but hopefully not drive to whatever might be still going on)

**Comment by Jailander** [comment]
I think it should preempt the current nav goal and put the robot in freerun

**Comment by Jailander** [comment]
maybe a simple version of the nogo stop script? https://github.com/strands-project/strands_navigation/blob/indigo-devel/emergency_behaviours/scripts/nogo_node_stop.py#L222

**Comment by marc-hanheide** [comment]
What I intended to do: 
* A small button directly provided by `strands_webserver` and therefore present on *any* screen, I would make it small and a little transparent but on top of anything else, alternative, we could block some small section at the top or bottom and reduce the size of the main content by let's say 60 pixels in width or height.
* Implement a node that will continuously call the service to engage freerun for about 30 seconds at a high frequency after the button is pressed.

Icon would be this:
![image](https://cloud.githubusercontent.com/assets/1153084/20829876/ef64671a-b875-11e6-9d02-31efdf096ff7.png)

**Comment by capamat-aaf** [comment]
Question regarding Bellbot:
I think @Jailander told me, but I forgot. Why are the buttons & typing in field arranged the way they are?
I think I'd much more used to a field for entering the address that spans over the screen and the two buttons next to the keyboard.

**Comment by marc-hanheide** [comment]
@capamat-aaf I did the layout ;-) So, could you provide a mockup... Then I could see if easily possible.

**Comment by capamat-aaf** [comment]
Thanks! Can you provide a screenshot? I can only find cut out images

**Comment by capamat-aaf** [comment]
I'm going to use this issue to write down some of the complaints:
- Trouble with e-wheelchairs: evasion was too spontaneous (n=2);
   • one wheelchairs apparently had a headrest in form of a long metal bar, he then bumped against it and got a bit stuck;
- People mentioned a general high spontaneity when evading (n=a couple of times)
- Leaving too quickly when people are still near him (n=a couple of times)
   • I saw this "happen" once; people were indeed in his proximity, but inactive regarding his usage
- It occurred that he stop too closely to somebody and when this person turned around, s/he almost tripped over him and fell.
- people told me that they jump out of his way because they don't trust his evasion manoeuvres

**Comment by marc-hanheide** [comment]
well... @cdondrup should read this as motivation ;-)
Some things there are just not possible: 
* a long metal bar is futile, no chance to see
* what do they mean with "spontaneity"?
* Leaving when people are near... Yes, not to be changed. It notices when people use (click) and won't wander off then, but if they just linger around, we don't model that.
* The getting close from behind is a well-known problem. Same happened to me at AAF when a cleaner's cart was left right behind me and I didn't notice it. It's a general awareness problem for people. We could beep continuously...
* the last jumping thing... yes, perceived safety is a very active reseach issue...

**Comment by capamat-aaf** [comment]
Yes, I'm also not quite sure how to approach the complaints.
I still hear about the toes incident and that happened some time ago...

By spontaneity I mean too last minute or too close to them; too sudden. 

I don't know if there is something that can be done for these two months; if a bigger halt-radius (?) or reducing velocity a bit would be doable or desired

**Comment by capamat-aaf** [comment]
![unnamed](https://cloud.githubusercontent.com/assets/21201058/22364269/8c302216-e470-11e6-8d9f-9c5b9c27280c.jpg)


This is a photo Veronika received from one of the directors. 
I don't know the content but it seem that he simply didn't stop.

**Comment by marc-hanheide** [comment]
OK, hard to tell the full story from just this photo. Clearly, the robot would not see the foot, but the leg. So, yes, this is a situation that can happen. Maybe, @cdondrup, it's time to put your code to use now? Or, with the current Human-aware navigation to be much more conservative and simply not move at all in the presence of humans... it will, however, cause a deterioration of the level of service and availability.

**Comment by cdondrup** [comment]
I didn't have time to do anything, yet. But I planned on starting to integrate it beginning of February.

Currently, the human-aware navigation works in a way that there is a cone in front of the robot where it considers humans and you can define a min distance at which it stops and a max distance at which it starts to slow down. Can mess with those as well.

**Comment by capamat-aaf** [comment]
Thank you for replying.
Is there anything that can be done so that I can pacify the spirits? Have him roam a bit slower or anything else?
It is indeed though to assess the situation, if I don't see it happen.
What I personally noticed once was, and perhaps I didn't explain it too well, that there are "problems" when he isn't going straight.
When approaching a node, we will at some point turn to it. This behaviour can be quite abrupt and inconvenience anyone who is on the same path.
![untitled](https://cloud.githubusercontent.com/assets/21201058/22365028/9c5c42b4-e475-11e6-9c32-fa37c8efb495.jpg)

And if both people are walking at normal speed, the person actually has to evade Henry.
I saw this happen at the café, where the person had to jump into the café (after the wall ends) to not bump into Henry.
I guess Henry would've stopped, but I would actually inquire maybe adapting his behaviour when turning, but this is just how I explained it to myself.

**Comment by capamat-aaf** [comment]
I talked to the person who observed the incident above and took the picture. 
He said that Henry would not stop.
I assume that Henry thought he wouldn't have to, since he thought he can pass between the person and the wall and then obviously not noticed the foot. 

In cases like this, where he decides to pass through intro a narrow path between people and other people/objects/walls, would it be possible for him to rather stop and wait?
When this was the case for eg cleaning carts or objects, his decision to still go on worked out very well, as far as I witnessed.

**Comment by marc-hanheide** [comment]
We can (as @cdondrup said) increase the "carefulness" around people. It would mean, however, that it  won't be able to accomplish its task for often. @cdondrup any advice what to change?

**Comment by cdondrup** [comment]
The config parameters can be found here: https://github.com/strands-project/strands_hri/blob/hydro-devel/strands_human_aware_navigation/cfg/HumanAwareNavigation.cfg

You can change the `detection_angle` which currently is at 45 degrees whish means 22.5 degrees to either side of the front of the robot. If increased it will also take detection into account that are not directly in front of it.

If you decrease the `min_dist` it will stop earlier. The speed is gradually reduced based on the distance of the person:

```
0.55m/s if dist >= max_dist
0.0m/s if dist <= min_dist
```

Everything in between makes it go slower. So increasing the min_dist makes it more cautious but in combination with increasing the angle will also result in situations where the robot will stop when someone goes past it.

**Comment by capamat-aaf** [comment]
Thank you @cdondrup 
It seems that there is some leeway to try things out. Perhaps we won't have to decrease the speed in this case and potentially destroy data @marc-hanheide @Jailander @gestom

**Comment by Jailander** [comment]
@cdondrup silly question when you said this
> If you decrease the min_dist it will stop earlier. The speed is gradually reduced based on the distance of the person:

you meant increase?

**Comment by cdondrup** [comment]
Of course I did. Isn't that obvious ;)

Thanks for pointing that out. I meant increase of course.

**Comment by marc-hanheide** [comment]
Do you will comfortable to make some recommendations for new numbers then?

**Comment by cdondrup** [comment]
Actually, I set it quite low during the walking group, e.g. `min_dist = 0` so it would never stop to prevent frustrating the therapists. It entirely depends you. I wouldn't necessarily change the `max_dist` but you can try increasing the `min_dist` to something like 2 metres (if it isn't there already) and the `detection_angle` to 90 degrees. If that is too conservative and the robot doesn't reach its goals anymore, try reducing the angle.
